src/client/clientmap.cpp

```diff
@@ -299,213 +291,291 @@ void ClientMap::updateDrawList()
-	v3s16 camera_cell = mesh_grid.getCellPos(camera_block);
+	/*
+	 When range_all is enabled, enumerate all blocks visible in the
+	 frustum and display them.
+	 */
+	if (m_control.range_all) {
+		MapBlockVect sectorblocks;
 
-	// Bits per block:
-	// [ visited | 0 | 0 | 0 | 0 | Z visible | Y visible | X visible ]
-	MapBlockFlags meshes_seen(mesh_grid.getCellPos(p_blocks_min), mesh_grid.getCellPos(p_blocks_max) + 1);
+		for (auto &sector_it : m_sectors) {
+			MapSector *sector = sector_it.second;
+			if (!sector)
+				continue;
 
-	// Start breadth-first search with the block the camera is in
-	blocks_to_consider.push(camera_mesh);
-	meshes_seen.getChunk(camera_cell).getBits(camera_cell) = 0x07; // mark all sides as visible
+			sectorblocks.clear();
+			sector->getBlocks(sectorblocks);
 
-	std::set<v3s16> shortlist;
+			// Loop through blocks in sector
+			for (MapBlock *block : sectorblocks) {
+				MapBlockMesh *mesh = block->mesh;
 
-	// Recursively walk the space and pick mapblocks for drawing
-	while (blocks_to_consider.size() > 0) {
+				// Calculate the coordinates for range and frustum culling
+				v3f mesh_sphere_center;
+				f32 mesh_sphere_radius;
 
-		v3s16 block_coord = blocks_to_consider.front();
-		blocks_to_consider.pop();
+				v3s16 block_pos_nodes = block->getPos() * MAP_BLOCKSIZE;
 
-		v3s16 cell_coord = mesh_grid.getCellPos(block_coord);
-		auto &flags = meshes_seen.getChunk(cell_coord).getBits(cell_coord);
+				if (mesh) {
+					mesh_sphere_center = intToFloat(block_pos_nodes, BS)
+							+ mesh->getBoundingSphereCenter();
+					mesh_sphere_radius = mesh->getBoundingRadius();
+				} else {
+					mesh_sphere_center = intToFloat(block_pos_nodes, BS)
+							+ v3f((MAP_BLOCKSIZE * 0.5f - 0.5f) * BS);
+					mesh_sphere_radius = 0.0f;
+				}
 
-		// Only visit each block once (it may have been queued up to three times)
-		if ((flags & 0x80) == 0x80)
-			continue;
-		flags |= 0x80;
+				// Frustum culling
+				// Only do coarse culling here, to account for fast camera movement.
+				// This is needed because this function is not called every frame.
+				float frustum_cull_extra_radius = 300.0f;
+				if (is_frustum_culled(mesh_sphere_center,
+						mesh_sphere_radius + frustum_cull_extra_radius)) {
+					blocks_frustum_culled++;
+					continue;
+				}
 
-		blocks_visited++;
+				// Raytraced occlusion culling - send rays from the camera to the block's corners
+				if (occlusion_culling_enabled && mesh
+						&& isMeshOccluded(block, mesh_grid.cell_size, cam_pos_nodes)) {
+					blocks_occlusion_culled++;
+					continue;
+				}
 
-		// Get the sector, block and mesh
-		MapSector *sector = this->getSectorNoGenerate(v2s16(block_coord.X, block_coord.Z));
+				if (mesh_grid.cell_size > 1) {
+					// Block meshes are stored in the corner block of a chunk
+					// (where all coordinate are divisible by the chunk size)
+					// Add them to the de-dup set.
+					shortlist.emplace(mesh_grid.getMeshPos(block->getPos()));
```

Comment:
This emplace essentially overrides isMeshOccluded() above. The best way would be to collect all meshes based on what's loaded + frustum and then call isMeshOccluded on what's in the short list.

The other algorithm is different because it only checks blocks with meshes.

Reply:
I guess that explains why I didn't notice an improvement when I changed from isBlockOccluded to isMeshOccluded.

In then end, this is for the case where a user wants to see all blocks, and most time will be spent in rendering all those blocks, so I'd like to not complicate this more than needed.

Reply:
Actually, are you sure isMeshOccluded actually works? I never see any blocks occluded in the debug view (F6), even as I move around or move behind a "mountain".

Reply:
BTW. I tried what you suggested (calling isMeshOccluded on the shortlisted blocks), but then I get missing blocks on the client.

Reply:
Oh I see what you are saying. We may occlusion cull a mesh, but then another block from the grid brings the mesh-block back in. T

Reply:
OK... Last comment... Since in this view we truly want to see all blocks loaded at the client, we also just forgo the occlusion culling.

I tried to do it for each block (mesh or not) and that is too slow.

What do you think, just not do occlusion culling in this case?